src/cargo/core/workspace.rs

```diff
@@ -538,14 +538,26 @@ impl<'cfg> Workspace<'cfg> {
             None
         };
 
-        for path in members_paths {
+        for path in &members_paths {
             self.find_path_deps(&path.join("Cargo.toml"), &root_manifest_path, false)?;
         }
 
         if let Some(default) = default_members_paths {
             for path in default {
-                let manifest_path = paths::normalize_path(&path.join("Cargo.toml"));
+                let normalized_path = paths::normalize_path(&path);
+                let manifest_path = normalized_path.join("Cargo.toml");
                 if !self.members.contains(&manifest_path) {
+                    // default-members are allowed to be excluded, but they
+                    // still must be referred to by the original (unfiltered)
+                    // members list. Note that we aren't testing against the
+                    // manifest path, both because `members_paths` doesn't
+                    // include `/Cargo.toml`, and because excluded paths may not
+                    // be crates.
+                    let exclude = members_paths.contains(&normalized_path)
```

Comment:
How come this condition is here as well? I may not be fully remembering what the `members_paths` set is either...

Reply:
`members_paths` is the glob-expanded paths of `members` (without `Cargo.toml`s). Without this check I believe the test (Which this patch adds) `excluded_default_members_still_must_be_members` failed -- e.g. you could have `members = ["foo"]`, `default-members = ["foo", "bar"]` and `exclude = ["bar"]`.